File: algorithms/go/312_maxCoins.go
package algorithms

func maxCoins(nums []int) int {
	var newNums []int
	newNums = append(newNums, 1)
	newNums = append(newNums, nums...)
	newNums = append(newNums, 1)
	nums = newNums

	dp := make([][]int, len(nums))
	for i := 0; i < len(dp); i++ {
		dp[i] = make([]int, len(nums))
	}

	for l := 2; l < len(nums); l++ {
		for i := 0; i < len(nums)-l; i++ {
			j := i + l
			for k := i + 1; k < j; k++ {
				dp[i][j] = maxAny(dp[i][j], nums[i]*nums[k]*nums[j]+dp[i][k]+dp[k][j])
			}
		}
	}

	return dp[0][len(nums)-1]
}

File: algorithms/c/312_maxCoins.c
//
// Created by wade on 2019/10/8.
//

#include "common.h"

// TODO: 二维动态规划
int maxCoins(int *nums, int numsSize) {

    int **dp;

    numsSize += 2;

    dp = malloc(sizeof(int) * numsSize);
    for (int i = 0; i < numsSize; i++) {
        dp[i] = malloc(sizeof(int) * numsSize);
    }

    for (int len = 2; len < numsSize; ++len) {
        for (int i = 0; i < numsSize - len; ++i) {
            int j = i + len;
            for (int k = i + 1; k < j; ++k) {
                dp[i][j] = MAX(dp[i][j], nums[i] * nums[k] * nums[j] + dp[i][k] + dp[k][j]);
            }
        }
    }
    return dp[0][numsSize - 1];
}
